### PR TITLE
Site Editor: Add Dropdown to Create Generic Templates

### PR DESCRIPTION
**Pull request by Copons**
## Description

Part of #26270

Add a very basic interface to create generic templates from the Site Editor navigation sidebar.
Clicking on the `+` button will open a dropdown listing all _missing_ [generic templates](https://github.com/WordPress/gutenberg/blob/e2d8b7a2c544d1142e7bfbbb1422a609d78e050d/packages/edit-site/src/utils/get-template-info/constants.js), filling them with the hierarchically closest template content, plus a "General" template that creates a blank template.

cc @jameskoster (for some reasons I can't add you to the reviewers)

### Major Notes

1. New templates are now `draft`. The idea is to avoid having them immediately applied to the site, while they might still be WIP.
Though, we don't have a proper way to work with draft templates (e.g. the "save" button is only "update", there is no publish flow).
2. We'll probably eventually need to have "fixed" slugs for `$custom` templates, otherwise there might be the risk that the user accidentally creates one with a "reserved" name (e.g. "Front Page" which would be slugified into `front-page`).

## How has this been tested?

- Enable the Full Site Editing experiment.
- Open the Site Editor.
- Click on the WP/site icon button, and then on "Templates".
- Click on the `+` button near the Templates menu title.
- Make sure a dropdown with a list of generic templates shows up.
- Make sure the templates already available on the site are not listed.
- Click on one of the "named" templates.
- Make sure the editor content is replaced with the appropriate template's content (e.g. `archive` should insert the `index` template content), and the document title shows the new template title.
- Click on the "General" template.
- Make sure the editor content is empty, and the document title is "New Template".

## Screenshots 

Screenshot 2020-10-23 at 12 50 26

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jameskoster** [comment]
Let's try using a popover rather than a modal:



Any existing templates should be hidden rather than faded out. "General" template == `$custom` template.

Clicking a template takes you to the site editor and populates the Canvas with the appropriate blocks. To determine the appropriate blocks we can hopefully reference the closest related template hierarchically. E.G. if I want to add the Page template, look first for the Singular template, then the Index template. Whichever is found is loaded as the initial state.

**Comment by Copons** [comment]
@jameskoster I've updated this with I think most of your suggestions.
The description has been updated as well with new discussion points and new testing instructions.

**Comment by david-szabo97** [comment]
> We probably need to have "fixed" slugs for $custom templates, otherwise there might be the risk that the user accidentally creates one with a "reserved" name (e.g. "Front Page" which would be slugified into front-page).

Can't we use a prefix? `gtnbrg_custom_$custom` or something? (Note I used underscore instead of dash)

**Comment by mtias** [comment]
Let's skip custom templates initially and focus on the supported ones.

**Comment by david-szabo97** [comment]
> New templates are now draft. The idea is to avoid having them immediately applied to the site, while they might still be WIP.

Shouldn't they be `auto-draft`? That's what we have in this [issue](https://github.com/WordPress/gutenberg/issues/25868).

**Comment by jameskoster** [comment]
Couple of small tweaks – we can probably omit singular for now, Single + Page cover the vast majority of use cases and eliminate some potential confusion.

For the template descriptions, I think we can still do some fine-tuning there. I'm inclined to say they should be literal and illustrate when they are applied. Let's use this order too:

Archive
Applied to archives like your posts page, categories, or tags

Single
Applies to individual content like blog posts

Page
Applies to pages

Search results
Applies to search results

404
Applies when content cannot be found

Default (Index)
Applies with no other template is found

**Comment by mtias** [comment]
@jameskoster I'd define Index more as _Main template, applying when no other template is found_. It's still considered the only file a theme needs.

**Comment by Copons** [comment]
@jameskoster @mtias the "generic" templates names and descriptions are all in this file: https://github.com/WordPress/gutenberg/blob/e2d8b7a2c544d1142e7bfbbb1422a609d78e050d/packages/edit-site/src/utils/get-template-info/constants.js

I think fine tuning their copy is not essential for this PR, so I'd rather have a separate one just for that. 🙂 

> Shouldn't they be `auto-draft`? That's what we have in this [issue](https://github.com/WordPress/gutenberg/issues/25868). This is confusing 🤔

Great question, and I don't have a good answer!
Teorethically, `auto-draft` is supposed to be assigned to file templates converted to `wp_templates` and yet untouched by the user.
Either way, if in this PR I try to create a template with status `auto-draft` via the REST API, I get an error 400: "status is not one of publish, future, draft, pending, private." 🤔

**Comment by jameskoster** [comment]
There have been a few discussions on Slack around how to handle the somewhat confusing overlap between the singular, single, and page templates. I feel like we're very close to a decision so would invite any further feedback on the following before merging.

Here are the two options that are currently being considered:



FWIW My preference is option 2 as it feels more intuitive, and I can see it elegantly scaling as CPTs are registered (see 2a). Due to the ambiguous resolution and associated complexities of the single and singular templates, I think we may be able to leave them out for now, and add them later via an affordance in the UI that enables folks to add more advanced templates.

That said, I am totally happy to push forward with option 1 as well. I know this is a bit in the weeds for where we are with this right now :)

**Comment by vindl** [comment]
> There have been a few discussions on Slack around how to handle the somewhat confusing overlap between the singular, single, and page templates. I feel like we're very close to a decision so would invite any further feedback on the following before merging.

I agree with @Copons that this is not essential for this PR and can be updated in a follow up. I wouldn't consider it a blocker for merging.

**Comment by vindl** [comment]
> 2. General templates ($custom) are created as "New Template", but we don't have a way in the Site Editor to rename them. Document settings sounds like a good place to me. 🤔
> 3. We probably need to have "fixed" slugs for $custom templates, otherwise there might be the risk that the user accidentally creates one with a "reserved" name (e.g. "Front Page" which would be slugified into front-page).

Both of this can be avoided for now if we disable custom template creation for the time being. I think it's better if we deal with that in a separate PR that can properly address renaming in Document Settings dropdown too. Having default templates will probably be enough to unblock initial testing.

File: packages/edit-site/src/components/navigation-sidebar/navigation-panel/new-template.js
Comment (vindl) [inline packages/edit-site/src/components/navigation-sidebar/navigation-panel/new-template.js:83]: I'd be fine with shipping this as an initial version after this menu group is removed.

**Comment by Copons** [comment]
I've rebased after the Nav Search feature, which brings in some button styling, and allowed me to add a "title action" container.
I've also removed the `$custom` template from the dropdown.

**Review by vindl — APPROVED** [review]
LGTM! :shipit: Nice work @Copons!

**Pull request merged**
